semmc/tests/semstore/ParamFormulaTests.hs

```diff
@@ -0,0 +1,316 @@
+      property $
+      E.handleAll (\e -> annotate (show e) >> failure) $ do
+        Some r <- liftIO newIONonceGenerator
+        CBO.withYicesOnlineBackend @(CBO.Flags CBO.FloatReal) r CBO.NoUnsatFeatures $ \sym -> do
+          -- generate a formula
+          let opcode = OpPack
+          (p, operands) <- forAllT (genParameterizedFormula sym opcode)
+          -- ensure that formula compares as equivalent to itself
+          compareParameterizedFormulasSymbolically sym operands 1 p p
+          -- now print the formula to a text string
+          debugPrint $ "parameterizedFormula: " <> show p
+          debugPrint $ "# literalVars: " <> show (MapF.size $ SF.pfLiteralVars p)
+          debugPrint $ "# defs: " <> show (MapF.size $ SF.pfDefs p)
+          let printedFormula = FO.printParameterizedFormula (HR.typeRepr opcode) p
+          debugPrint $ "printedFormula: " <> show printedFormula
+          -- convert the printed text string back into a formula
+          let fenv = undefined
+          lcfg <- liftIO $ Log.mkLogCfg "rndtrip"
+          reForm <- liftIO $
+                    Log.withLogCfg lcfg $
+                    FI.readFormula sym fenv (HR.typeRepr opcode) printedFormula
+          debugPrint $ "re-Formulized: " <> show reForm
+          f <- evalEither reForm
+          -- verify the recreated formula matches the original
+          compareParameterizedFormulasSymbolically sym operands 1 p f
+
+    , testProperty "serialized formula round trip, online backend, OpSolo" $
+      property $
+      E.handleAll (\e -> annotate (show e) >> failure) $ do
+        Some r <- liftIO newIONonceGenerator
+        CBO.withYicesOnlineBackend @(CBO.Flags CBO.FloatReal) r CBO.NoUnsatFeatures $ \sym -> do
+          -- generate a formula
+          let opcode = OpSolo
+          (p, operands) <- forAllT (genParameterizedFormula sym opcode)
+          -- ensure that formula compares as equivalent to itself
+          compareParameterizedFormulasSymbolically sym operands 1 p p
+          -- now print the formula to a text string
+          debugPrint $ "parameterizedFormula: " <> show p
+          debugPrint $ "# literalVars: " <> show (MapF.size $ SF.pfLiteralVars p)
+          debugPrint $ "# defs: " <> show (MapF.size $ SF.pfDefs p)
+          let printedFormula = FO.printParameterizedFormula (HR.typeRepr opcode) p
+          debugPrint $ "printedFormula: " <> show printedFormula
+          -- convert the printed text string back into a formula
+          let fenv = undefined
+          lcfg <- liftIO $ Log.mkLogCfg "rndtrip"
+          reForm <- liftIO $
+                    Log.withLogCfg lcfg $
+                    FI.readFormula sym fenv (HR.typeRepr opcode) printedFormula
+          debugPrint $ "re-Formulized: " <> show reForm
+          f <- evalEither reForm
+          -- verify the recreated formula matches the original
+          compareParameterizedFormulasSymbolically sym operands 1 p f
+
```

Comment:
It is a bit hard to eyeball, but are these test bodies something that could be factored out into two or three shared functions? It is a bit hard to follow what is setup vs what is the main thing being tested - especially the three calls to `compareParameterizedFormulasSymbolically`.

Reply:
They can and should, and they mostly just need to be parameterized on opcode, which could then be another generated parameter, but the typing wasn't simple so I left this for later, but I think later is probably today.